### PR TITLE
IPN-21 remove concept of "sales cycle" IPTs

### PR DESCRIPTION
**Pull request by elmariachi111**
- Tokenizer indexes IPTs by IPNFT ids
- drops "hashes" completely
- Tokenizer:reinit adds existing IPTs to the known ipt mapping
- IPToken asks its owner (Tokenizer) if senders hold the underlying IPNFT
- even is upgradeable on top of the previous 1.1 -> 1.2 chain state
- loosens SalesShareDistributor's cap rules since there's only 1 ipt instance
- updates test cases where necessary

### CODE REVIEW

File: src/SalesShareDistributor.sol [hunk 1]
```diff
@@ -85,19 +91,20 @@ contract SalesShareDistributor is UUPSUpgradeable, OwnableUpgradeable, Reentranc
...
      *         this is a deep dependency on our own sales contract
      *
      * @param tokenContract IPToken     the tokenContract of the IPToken
      * @param listingId     uint256     the listing id on Schmackoswap
      * @param permissioner  IPermissioner   the permissioner that permits claims
      */
     function afterSale(IPToken tokenContract, uint256 listingId, IPermissioner permissioner) external {
-        if (_msgSender() != tokenContract.issuer()) {
+        Metadata memory metadata = tokenContract.metadata();
+        //todo: this should be the *former* holder of the IPNFT, not the 1st owner :)
```
Comment (nour-karoui) [inline src/SalesShareDistributor.sol:103]: we mean here the current owner of the IPNFT?

File: src/IPToken.sol [hunk 2]
```diff
@@ -13,75 +14,67 @@ struct Metadata {
...
 
-    function initialize(string calldata name, string calldata symbol, Metadata calldata metadata_) external initializer {
+    function initialize(uint256 ipnftId, string calldata name, string calldata symbol, address originalOwner, string memory agreementCid)
+        external
+        initializer
+    {
         __Ownable_init();
         __ERC20_init(name, symbol);
-        _metadata = metadata_;
+        _metadata = Metadata({ ipnftId: ipnftId, originalOwner: originalOwner, agreementCid: agreementCid });
```
Comment (nour-karoui) [inline src/IPToken.sol:43]: shouldn't we have the originalIpmftOwner just like we have the ipnftId?

**Review by nour-karoui — COMMENTED** [review]
After implementing the commented todos (//todos), will we also remove the originalOwner in the metadata? because I don't understand why we need to save it?

File: src/IPToken.sol (hunk 2 above)
Comment (elmariachi111) [reply src/IPToken.sol:43]: ?

File: src/IPToken.sol [hunk 3]
```diff
@@ -13,75 +14,67 @@ struct Metadata {
...
+        }
         totalIssued += amount;
         _mint(receiver, amount);
     }
 
     /**
      * @notice mark this token as capped. After calling this, no new tokens can be `issue`d
      */
-    function cap() external onlyIssuerOrOwner {
+    function cap() external onlyTokenizerOrIPNFTHolder {
```
[the remark this answers is not in the document]
Comment (elmariachi111) [reply src/IPToken.sol:77]: the caller can check this.

File: src/IPToken.sol (hunk 2 above)
Comment (nour-karoui) [reply src/IPToken.sol:43]: I meant the naming: we name originalIpnftOwner instead of originalOwner

File: src/SalesShareDistributor.sol (hunk 1 above)
Comment (elmariachi111) [reply src/SalesShareDistributor.sol:103]: this contract is not used. When there are todos in a contract codebase we should consider it not production ready. The idea of this concept is that the *previous* owner of an IPNFT receives sales proceedings and distributes them among the former holders.

File: src/SalesShareDistributor.sol [hunk 4]
```diff
@@ -85,19 +91,20 @@ contract SalesShareDistributor is UUPSUpgradeable, OwnableUpgradeable, Reentranc
 
     /**
      * @notice anyone should be able to call this function after having observed the sale
-     *         rn we restrict it to the token issuer since they must provide a permissioner that controls the claiming rules
+     *         rn we restrict it to the original owner since they must provide a permissioner that controls the claiming rules
```
[the remark this answers is not in the document]
Comment (elmariachi111) [reply src/SalesShareDistributor.sol]: changed it to the trade listing creator aka the seller.

**Comment by elmariachi111** [comment]
the `originalOwner` property is part of the storage layout of the permissioner contract that can't be upgraded. Hence we cannot remove it without breaking the already existing contracts. (see Permissioner.sol:63 -> the old IPTs use that memory layout. Even if I change it on IPToken, it will still be laid out like that)

File: test/SalesShareDistributor.t.sol [hunk 5]
```diff
@@ -160,17 +162,17 @@ contract SalesShareDistributorTest is Test {
         vm.startPrank(ipnftBuyer);
         erc20.transfer(originalOwner, 1_000_000 ether);
 
-        // only the owner can manually start the claiming phase.
+        // only the origina owner can atm  manually start the claiming phase.
```
Comment (nour-karoui) [inline test/SalesShareDistributor.t.sol:165]: typo: original

File: src/Tokenizer.sol [hunk 6]
```diff
@@ -69,7 +70,7 @@ contract Tokenizer is UUPSUpgradeable, OwnableUpgradeable {
     }
 
     //todo: try breaking this with a faked IPToken
-    modifier onlyControlledIPTs(IPToken ipToken) {
+    modifier onlyController(IPToken ipToken) {
         TokenMetadata memory metadata = ipToken.metadata();
 
         if (address(synthesized[metadata.ipnftId]) != address(ipToken)) {
```
Comment (nour-karoui) [inline src/Tokenizer.sol:76]: shouldn't we rename the error to IPNotControlledByController

**Review by nour-karoui — APPROVED** [review]

**Pull request merged**